### ISSUE TITLE
hb_blob_get_data not working on Windows since version 4.4

### ISSUE THREAD

**Issue by liZe** [harfbuzz/harfbuzz#3752]
Hi!

First of all, thanks for this marvelous library 🫶🏽.

We use Harfbuzz for WeasyPrint, a web engine in Python for document generation, and since Harfbuzz version 4.4 has been released my CI fails on Windows. Everything was OK with 4.3 for all OSes, and everything is OK with 4.4 on Linux.

My current code takes the Harfbuzz font out of a Pango font and tries to read the blob associated to the font face. Here is the pseudo-code:

```
hb_font = pango_font_get_hb_font(pango_font)
hb_face = hb_font_get_face(hb_font)
hb_blob = hb_face_reference_blob(hb_face)
hb_data = hb_blob_get_data(hb_blob, length)
```

Since 4.4.1 (and maybe 4.4.0, I can’t test), `hb_data` is a null pointer (and length is 0). `hb_font`, `hb_face` and `hb_blob` seem to be valid pointers (at least they’re not null).

I’ve read the changelog but didn’t find anything obviously related to this problem. I couldn’t find a bug in the bug tracker either. Do you have any idea about where this problem could come from?

I’ll be happy to provide more information if needed.

(I also posted this message on the mailing list, but there will probably be more activity here.)

**Comment by behdad** [comment]
Thanks for the kind words. As you note, we didn't change anything in that respect.

Is it possible that you are using a Pango built with DirectWrite instead of Uniscribe or FreeType? I think that would cause something like this.

**Comment by liZe** [comment]
Thanks for the answer.

[Harfbuzz is built using DirectWrite](https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-harfbuzz/PKGBUILD#L59), but I’m not sure about Pango. Is [this disabled xft option](https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-pango/PKGBUILD#L51) related to the problem?

**Comment by behdad** [comment]
> [Harfbuzz is built using DirectWrite](https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-harfbuzz/PKGBUILD#L59),

Okay I'm guessing that it has to do with directwrite then.

Looks like there's API to get the blob itself: https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nf-dwrite-idwritefontfilestream-readfilefragment
but that's on the stream interface, and I don't see a way to get the stream from a face :(. I'll keep poking.

**Comment by liZe** [comment]
> I'll keep poking.

Thanks.

But the original mystery remains: why did version 4.3.0 work correctly? The build options are exactly the same (here’s the [history](https://github.com/msys2/MINGW-packages/commits/master/mingw-w64-harfbuzz/PKGBUILD)). I’ll continue investigating…

**Comment by behdad** [comment]
Yeah that is mysterious. I have no idea.

**Comment by liZe** [comment]
I can confirm that the problem comes from a change in the code between 4.3.0 and 4.4.1. I’ve tried both versions with the exact same compilation options on the exact same environment, <=4.3.0 works and >=4.4.1 fails (I’ve tried 5.1.0 too).

As I use GitHub Actions and can’t easily use a Windows machine, it will be a nightmare to find the source of the problem. I’d love to use my beloved `git bisect`… Any tips to print useful debug messages from Harfbuzz at runtime?

**Comment by behdad** [comment]
> Any tips to print useful debug messages from Harfbuzz at runtime?

Try building with `HB_DEBUG_BLOB` or `HB_DEBUG` or `HB_DEBUG=100`. The first one might be enough for this problem.

**Comment by liZe** [comment]
> Try building with `HB_DEBUG_BLOB` or `HB_DEBUG` or `HB_DEBUG=100`. The first one might be enough for this problem.

Well… I use MinGW’s pre-built packages on GitHub Actions, I can’t rebuild Harfbuzz easily. Looks like I’ll have to find a Windows machine to try this :/.

**Comment by khaledhosny** [comment]
Can you check that the font, face, and blob you get are not empty (e.g `hb_font != hb_font_get_empty()`?

**Comment by liZe** [comment]
> Can you check that the font, face, and blob you get are not empty (e.g `hb_font != hb_font_get_empty()`?

Using the latest version fixes the problem. The fix happened somewhere between 5.1.0 and 7.1.0, I don’t know when exactly.

Thanks for your patience.

**Comment by behdad** [comment]
That's truly weird.

That said, I still don't see a way to get the full blob from DirectWrite.
